### PR TITLE
[JENKINS-63654] X-Correlation-ID to monitor Jenkins landscapes called remotly

### PR DESCRIPTION
**Pull request by llecaroz**
In complex automated build systems cascading events and starting builds
remotely, it can be useful to propagated a X-Correlation-ID to be able
to map related events all together from the begining to the end

As the initial piece of code is there, now, for example, others clients
starting builds remotely can propagate correlation Id which can be
retrieve thru the getCorrelationId() and on their turn, send
it/cascade/propagate it later on.

By using the Run.getCause(Cause.class).getCorrelationId, people will be able to understand from which remote system/request, this build was generated from & will be able to propagate it to others sub calls(others remote jobs, microservices on the backend or whatever), & then to follow from the begining/first caller having created the job to the end. 

For example:
- Parameterized Remote Trigger Plugin to send a X-Correlation-ID in the header to Jenkins & get it stored in the Cause
- SCM Triggers to receive events containing X-Correlation-ID & also store them in the Cause
- Next others plugins invoked (Build Steps) in a Job which need to know from which original sender this job was started & to send this information to others backing services used in Build Steps or others.

See [JENKINS-63654](https://issues.jenkins-ci.org/browse/JENKINS-63654).

### Proposed changelog entries

* correlationId getter & setter added in Cause.java to enable saving X-Correlation-ID stored in the header of the request (Cause.java)
* doBuild & doBuildWithParameters HTTP end points saving the X-Correlation-ID inside the created Cause

(as none of this class were previously tested & due to the very small changes proposed here, no unit test was proposed)

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/jenkins/model/ParameterizedJobMixIn.java [hunk 1]
```diff
@@ -260,6 +260,8 @@ public static final CauseAction getBuildCause(ParameterizedJob job, StaplerReque
         } else {
             cause = new Cause.UserIdCause();
         }
+        
+        cause.setCorrelationId(req!=null?req.getHeader("X-Correlation-ID"):null);
```
Comment (varyvol) [inline core/src/main/java/jenkins/model/ParameterizedJobMixIn.java:264]:
```suggestion
        cause.setCorrelationId(req != null ? req.getHeader("X-Correlation-ID"):null);
```

**Comment by alecharp** [comment]
I do understand the point of this PR and see the benefit it can provides to users. However, it feels like this PR is not enough on its own. Wouldn't it require a downstream PR in some plugins to "provide" the header? 

My question is more about: where is the header added in the request? Does it mean that the requests has to be generated by a CLI / tools which would add the header in the request?

Thanks.

**Comment by llecaroz** [comment]
Hi @alecharp , 

Yes, in fact, these are the first bricks which will be helpful & a pre-requisite to extend after having accepted that PR. 
There will be certainly more downstream PR, for example, I will certainly update  the Parameterized Remote Trigger plugin to accept a X-Correlatoin-ID as new parameter (set by a pipeline or whatever) or to reuse the one if already existing from the parent build which invoked the PRT plugin to propagate it to the next remote Jenkins server, etc...

So here in my example, the PRT plugin will check if a correlation id was given as paremeter or existing from the Job in which it is called, set it in the header of the /build or /buildWithParameters requests, next it will be retrieved as we can see here & set in the Cause, & if there is another PRT plugin called on that jenkins, it will again get it from the cause of the build inside the PRT plugin is invoked, & continue its propogation to others remote jenkins on which it needs to be cascaded.

After that, others plugins behind sending tcp requests to some microservice(, will be able to get the correlation id from the cause, & *it will be easy to track the originator* of the initial request along a complex landscape having jenkins servers in the middle.

Hope it answered to your point ? But for sure, some plugins will have to be adjusted to propagate it like the PRT plugin as said & certainly more we use behind the scene.

**Review by daniel-beck — CHANGES_REQUESTED** [review]
I'm unsure this is a reasonable addition; remotely triggered causes already support an optional message that can be used for something like this.

I also expect this would need propagation to downstream causes via builds triggering other builds etc. to be really useful.

Besides these concerns about the approach, this needs a lot more documentation, both in source code (Javadoc) and elsewhere. How would users even know this exists and can be used?

Also, complete lack of tests.

I would recommend you start with getting consensus about this addition in principle before spending time on implementation / documentation concerns.

**Comment by llecaroz** [comment]
Hi @daniel-beck , 

you are bringing some very relevant points. Initially, I tried to think in an agnostic way doing this automatically (for example the to see if some observers could be added at the jvm http/tcp level by using TLS, ...) (or Stapler framework) but this seems quite impossible from my knowledge even more because people can use their own comm API :(

& even if the Jenkins Object Model is adjusted to internally manage automatically a such propagation mechanism, this seems also to be complex because, as said, people could use their own comms APIs at the end or in some parts of the chain.

**Do you know if there is a forum or a place in which a such topic could be shared & put in place ?**

Due to the complexity of the topic, I tried to be as most as possible, pragmatic, meaning, trying to follow requests from end to end  when  some building cascading events are sent along a full CI/CD chain.

Like the PRT , if this PR is accepted, I will certainly in the future also propose some updates, not only in the PRT plugin but in GIT/Gerrit/Github triggering plugins which could help into troubleshooting transversal landscapes issues.

Looking forward to reading from you
Best regads

**Comment by daniel-beck** [comment]
> **Do you know if there is a forum or a place in which a such topic could be shared & put in place or elaborated ?**
> I think that this is a very important & relevant topic looking to the future & I will be very happy to contribute on a such topic

Look into https://github.com/jenkinsci/jep/ but note that the process is partially defunct at the moment. It's still the best framework for someone who's not a regular contributor to propose larger-scale changes and get consensus.

If this would end up being fairly small overall, a thread on the developers list should do it: https://www.jenkins.io/mailing-lists/

Otherwise, perhaps this maps well to one of the SIGs: https://www.jenkins.io/sigs/ (at a glance it doesn't, but mentioning just in case)

**Comment by llecaroz** [comment]
Thank you @daniel-beck , I will open a corresponding discussion in the appropriate link you gave me.

> I'm unsure this is a reasonable addition; remotely triggered causes already support an optional message that can be used for something like this.

The problem is that sometimes even more like when there is no token, the Cause is instantiated instead of RemoteCause. If we also want triggering events, to propagate correlation ids, having it in Cause, will ensure that it will be faisable from others plugins to add/propagate correlations like in that one (just as example):  https://github.com/jenkinsci/generic-webhook-trigger-plugin/search?q=Cause&unscoped_q=Cause
As you can see, the problem is that not all triggering plugins even if they are webhook based doesn't inherite from RemoteCause

> I also expect this would need propagation to downstream causes via builds triggering other builds etc. to be really useful.

Hum.... good point: looking to here: https://github.com/jenkinsci/jenkins/blob/676d02a3b97848eb3a884e27862c7f6094037ac4/core/src/main/java/hudson/model/Cause.java#L194
Because the upstream can contain multiple causes, is it potentially possible that more than one contains multiple causes coming from remote requests ? So at the end the Run gets multiple correlation ids as source ? in a such situation, if this is possible, wil be hard to know which one to cascade even more if it has to be set in a HTTP header. If it only stays inside cause, this will be propagated automatically as the UpstreamCause will store upstream causes. but if it has to be sent to http requests....

**Comment by MRamonLeon** [comment]
Agree with Daniel on raising the discussion on the mailing list, taking into account the scope of these changes.

**Comment by alecharp** [comment]
@llecaroz I cannot find the discussion on the mailing list, did you start it? Do you plan to get back to this or should we close the PR? I'm adding the label `proposed-for-close` for now.

**Comment by basil** [comment]
Wow, this looks like a very interesting proposal! I think this could be a valuable addition to Jenkins.

Given the scope of this change, I suggest creating a [Jenkins Enhancement Proposal (JEP)](https://github.com/jenkinsci/jep) and sharing it on the [developer list](https://www.jenkins.io/mailing-lists/). This is a document that is used in the Jenkins Community in order to facilitate the design review process. The JEP should cite this PR as a reference implementation as well as providing examples of downstream consumers.

Due to the lack of design review and inactivity, I am closing this PR pending the creation of a JEP and examples of downstream consumers. Once design review has been completed, this PR can be reopened. On behalf of the core team, I would like to thank you for your contribution. Even though this PR has not yet made it across the finish line, it is a promising start! I continue to encourage you (or anyone else who is interested!) to pick up this effort and drive it to completion. Thanks!

**Pull request closed without merging**